### PR TITLE
Add separate repository container

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -61,7 +61,7 @@
             "common": "common.php",
             "console": "console.php",
             "events-console": "events-console.php",
-            "providers-web": "providers-web.php"
+            "providers": "providers-web.php"
```
Comment (samdark) [inline composer.json:64]: File should be renamed as well, I think.

Comment (yiiliveext) [reply composer.json:64]: The file is renamed, see below

Comment (samdark) [reply composer.json:64]: Yes but in this line it is still `"providers-web.php"`

Comment (yiiliveext) [reply composer.json]: Fixed

File: src/Factory/RepositoryContainer.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * @psalm-pure
+     * @param ORMInterface $orm
+     * @param string $role
+     * @return Closure
+     */
+    private function makeRepositoryFactory(ORMInterface $orm, string $role): Closure
+    {
+        return static fn (): RepositoryInterface => $orm->getRepository($role);
```
Comment (roxblnfk) [inline src/Factory/RepositoryContainer.php:82]: You don't need to create closures in the container.

File: src/Factory/RepositoryContainer.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        if ($this->has($id)) {
+            return $this->instances[$id] = $this->repositoryFactories[$id]();
+        }
+
+        throw new NotFoundException("Repository $id doesn't exist.");
+    }
+
+    public function has($id): bool
+    {
+        if (!$this->build) {
```
Comment (roxblnfk) [inline src/Factory/RepositoryContainer.php:47]: You should make container more lazy.
Before building do checking that $id is subclass of `RepositoryInterface`

**Review by roxblnfk — CHANGES_REQUESTED** [review]

File: src/Factory/RepositoryContainer.php (hunk 2 above)
Comment (yiiliveext) [reply src/Factory/RepositoryContainer.php:82]: You are right, this is just a draft, I wanted to check the idea.

File: src/Factory/RepositoryContainer.php (hunk 3 above)
Comment (yiiliveext) [reply src/Factory/RepositoryContainer.php:52]: Fixed

File: src/Factory/RepositoryContainer.php (hunk 2 above)
Comment (yiiliveext) [reply src/Factory/RepositoryContainer.php]: Fixed

File: src/Factory/RepositoryContainer.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        if (isset($this->instances[$id])) {
+            return $this->instances[$id];
+        }
+
+        if ($this->has($id)) {
+            return $this->instances[$id] = $this->makeRepository($this->roles[$id]);
+        }
+
+        throw new NotFoundException("Repository $id doesn't exist.");
```
Comment (roxblnfk) [inline src/Factory/RepositoryContainer.php:40]: It is preferable to throw several friendly errors here:
- "ambiguous repository definition" when the repository is used by several roles;
- "the repository is not defined in the ORM Schema" 
- "the requested object is not a repository (doesn't implements RepositoryInterface)"

File: src/Factory/RepositoryContainer.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                $this->roles[$repo] = current($role);
+            }
+        }
+    }
+
+    /**
+     * @psalm-pure
+     * @param ORMInterface $orm
+     * @param string $role
+     * @return Closure
```
Comment (roxblnfk) [inline src/Factory/RepositoryContainer.php:80]:
```suggestion
```

**Pull request closed without merging**